maintenance/wikia/Forum/fixCommentIndexes.php

```diff
@@ -0,0 +1,262 @@
+
+		foreach ( $threads as $threadId ) {
+			$brokenThread = new BrokenThread( $threadId );
+			$brokenThread->fix();
+		}
+	}
+
+	/**
+	 * Find the list of threads with children that have duplicated values in the props field
+	 * of page_wikia_props.  This indicates the counter has been reused by more than one comment
+	 * in the same thread.  We ignore deleted, removed and archived comments since those are likely
+	 * to have duplicates with current comments because the author of that code never considered
+	 * un-deleting/removing/archiving as a thing.  Not going to fix that here since who knows what
+	 * additional bugs that will uncover.
+	 *
+	 * @return array
+	 */
+	private function getAffectedThreads() {
+		$dbr = wfGetDB( DB_SLAVE );
+		$threads = ( new WikiaSQL )
+			->SELECT( 'parent_comment_id', 'props', 'count(*)' )
+			->FROM( 'comments_index' )
+			->JOIN( 'page_wikia_props' )
+			->WHERE( 'page_id' )->EQUAL_TO_FIELD( 'comment_id' )
+			->AND_( 'propname' )->EQUAL_TO( WPP_WALL_COUNT )
+			->AND_( 'parent_comment_id' )->NOT_EQUAL_TO( 0 )
+			->AND_( 'removed' )->EQUAL_TO( 0 )
+			->AND_( 'deleted' )->EQUAL_TO( 0 )
+			->AND_( 'archived' )->EQUAL_TO( 0 )
+			->GROUP_BY( 'parent_comment_id, props' )
+			->HAVING( 'count(*)' )->GREATER_THAN( 1 )
+			->runLoop( $dbr, function ( &$threads, $row ) {
+				$threads[] = $row->parent_comment_id;
+			} );
+
+		// Ensure that we're always sending an array back
+		return empty( $threads ) ? [ ] : $threads;
+	}
+
+	/**
+	 * Get the URL for the wiki being worked on by this script
+	 *
+	 * @return string
+	 */
+	static public function getWikiURL() {
+		$dbName = WikiFactory::IDtoDB( F::app()->wg->CityId );
+		$url = WikiFactory::DBtoDomain( $dbName );
+		return WikiFactory::getLocalEnvURL( 'http://'.$url );
+	}
+}
+
+class BrokenThread {
+
+	private $threadId;
+
+	/**
+	 * @param int $threadId
+	 */
+	public function __construct( $threadId ) {
+		$this->threadId = $threadId;
+	}
+
+	/**
+	 * Fix the comment numbering for the thread given by $this->threadId
+	 */
+	public function fix() {
+		FixCommentIndexes::debug( "* Fixing thread ".$this->threadId );
+
+		$lastIdx = $this->renumberComments();
+		$this->setNewWallCount( $lastIdx );
+	}
+
+	/**
+	 * Sets the new index to use whenever the next comment is made on the current thread
+	 *
+	 * @param int $idx Update the new last used index on this wall
+	 */
+	private function setNewWallCount( $idx ) {
+		if ( FixCommentIndexes::isTest() ) {
+			return;
+		}
+
+		wfSetWikiaPageProp( WPP_WALL_COUNT, $this->threadId, $idx );
+	}
+
+	/**
+	 * Update the comment number used for creating links to those comments.  It doesn't matter what they are
+	 * as long as each comment has a unique number, however ascending from 2 is how it works now and makes the
+	 * most sense.
+	 *
+	 * @return int The last index we assigned to a comment
+	 */
+	private function renumberComments() {
+		$comments = $this->getCommentIds();
```

Comment:
@kvas-damian its not that simple.  On save, new comments look at the `WPP_WALL_COUNT` for the parent.  It increments that value and writes it back to the parent and then updates itself: 
```php
$count = $parent->getOrderId( $userMaster = true );
if ( is_numeric( $count ) ) {
	$count++;
	$parent->setOrderId( $count );
	$class->setOrderId( $count );
}
```
(code here: https://github.com/Wikia/app/blob/b00dd8fce32b720af3fd646e45204effb7b9ed2e/extensions/wikia/Wall/WallMessage.class.php#L151-L156)

Note that even this wall code contains a race condition between the read of the master and the write of the new count that I have already seen in our older data.

So lets assume we have the following broken data for the parent and its comments shown, noting that comments always start at 2:

* parent count: 3
* comment 1 count: 2
* comment 2 count: 3
* comment 3 count: 3
* comment 4 count: 3

We have to fix *both* parent and comment counts.  Yah, we can follow your steps:

1. get comment
2. update it
3. go to first point

but until we update the parent's count, the next comment is always going to have the wrong count.  Here's the race condition:

T) Look and find a comment with a comment ID greater than the last comment we fixed
T+1) Update the comment count to $count += 1
T+2) Look and *fail* to find a comment ID greater than what we found in step N
T+3) _A comment is made in the system and takes the current parent count which is 3._
T+4) Now that we have failed to find any more comments in T+2, update parent count to $count

We now have a duplicate #3 count.  Looking at individual comments doesn't help because we always have a race condition in updating the parent count.  You can't do T+2 and T+4 atomically without locking the table.

Given the Wall code already lives with many bugs *and* there are obvious race conditions already *and* that we're replacing it *and* that the solution agreed up on by the community team accepts broken URLs as part of the fix, I'm not interested in spending a lot of time on a bulletproof solution for something already subpar.  We just need to fix the bulk of the problem.  If  1 wiki out of 1,000 experiences the race condition and leaves a duplicated comment behind, I'm ok with that.  Odds are that only 1 out 1,000 of those duplicates would actually be noticed by somebody, so for that 1 in a 1,000,000 situation, I'm ok re-running the script for that thread.